File: prj182_sharing.py
#!/usr/bin/python
#-*- coding:utf-8 -*-

#Author: Ben

def getLength(head, nodes):
    length = 0
    while head != '-1':
        length += 1
        head = nodes[head]
    return length

def getFirstCommonNode(head1, head2, nodes):
    l1 = getLength(head1, nodes); l2 = getLength(head2, nodes)
    if l1<l2:
        l1, l2 = l2, l1
        head1, head2 = head2, head1
    for i in range(l1 - l2):
        head1 = nodes[head1]
    rlt = '-1'
    for i in range(l2):
        if head1 == head2:
            rlt = head1
            break
        head1 = nodes[head1]
        head2 = nodes[head2]
    return rlt

while True:
    try:
        head1, head2, N = raw_input().strip().split()
        nodes = dict(); N = int(N)
        for i in range(N):
            first, ch, second = raw_input().strip().split()
            nodes[first] = second
        print getFirstCommonNode(head1, head2, nodes)
    except:
        break

File: prj182_sharing.cpp
#include <iostream>
#include <map>
#include <string>

using namespace std;

int getLength(string head, map<string, string> &nodes)
{
    int length = 0;
    while(head != "-1") {
        length++;
        head = nodes[head];
    }
    return length;
}

string getFirstCommonNode(string &head1, string &head2, map<string, string> &nodes)
{
    int l1 = getLength(head1, nodes), l2 = getLength(head2, nodes);
    if(l1 < l2) {
        swap(l1, l2); swap(head1, head2);
    }
    for(int i=0; i<(l1-l2); i++) {
        head1 = nodes[head1];
    }
    string rlt = "-1";
    for(int i=0; i<l2; i++) {
        if(head1 == head2) {
            rlt = head1; break;
        }
        head1 = nodes[head1];
        head2 = nodes[head2];
    }
    return rlt;
}

int main()
{
    ios::sync_with_stdio(false);
    string head1, head2; int N;
    while(cin >> head1 >> head2 >> N) {
        map<string, string> nodes;
        string first, second, ch;
        for(int i=0; i<N; i++) {
            cin >> first >> ch >> second;
            nodes[first] = second;
        }
        cout << getFirstCommonNode(head1, head2, nodes) << endl;
    }
    return 0;
}
